### ISSUE TITLE
Federated facebook login in react native requires two login attempts to work on IOS

### ISSUE THREAD

**Issue by erikgaas** [aws-amplify/amplify-js#7468]
**Describe the bug**
I am working through the tutorial for federated login here: https://docs.amplify.aws/lib/auth/social/q/platform/js#deploying-to-amplify-console

I am using the "In-App Browser Setup (optional, but recommended)" section to federate login into facebook from cognito.

This works fine on android with no trouble. The problem is testing it on the simulator for IOS. Following the tutorial based on auth command actually doesn't work. `Auth.federatedSignIn()` is the command the docs say to use. (Again this works fine in Android) If I use federatedSignIn without arguments it leads to this error every time I try to sign in.

```Sign in failure [ValidationException: 1 validation error detected: Value '{cognito-idp.us-west-2.amazonaws.com/us-west-2_rBVomGdCO=}' at 'logins' failed to satisfy constraint: Map value must satisfy constraint: [Member must have length less than or equal to 50000, Member must have length greater than or equal to 1]]```

Now if I change this to `Auth.federatedSignIn({provider: 'Facebook'})` then if I try signing in the first time, it will log the error above, but if I sign in again then everything works fine.

I'm pretty sure this is due to some weird mismatch in authentication state. It seems as though the first attempt to login initializes something important on the frontend that is able to send the correct request on the second auth call.

I have tried to hunt down the problem, presumably there is something I could to do inspect the auth state over time, but for now I've been looking at the requests through urlOpener. The only thing that is different is the `newUrl` returned by the InAppBrowser.openAuth call.

First call it is: appname:///?code=<uuid>&state=<state_id>#_=_
and second call is
appname://?code=<uuid>&state=<state_id>

So I guess the only difference is that #_=_ at the very end which makes zero sense to me. Given that the requests are the same to the backend and only the responses change it makes me think there is actually a backend state being mismanaged...???

**To Reproduce**

1. Try federated login with facebook using the recommended but changing federated login code to specify facebook as provider
2. Login once and observe error while being returned back to login screen
3. Try logging in again and have it succeed showing user information

Alternatively try not specifying provider and always see it fail

**Expected behavior**
I expect to be successfully authenticated such that Hub is able to listen for a successful authentication

**Code Snippet**
This is the crux of where the command is flaky. In federatedSignIn
```
  return (
    <View>
      <Text>User: {user ? JSON.stringify(user.attributes) : 'None'}</Text>
      {user ? (
        <Button title="Sign Out" onPress={() => Auth.signOut()} />
      ) : (
        <Button title="Federated Sign In" onPress={() => Auth.federatedSignIn({provider: 'Facebook'})} />
      )}
    </View>
  );
}
```

**Screenshots**
![image](https://user-images.githubusercontent.com/2773725/103375100-82816500-4a96-11eb-8057-a6fbd024db11.png)
Clicking on this once goes back to login page without populating user. Clicking on it twice populates user.

<details>
  <summary><strong>Environment</strong></summary>

```
npx envinfo --system --binaries --browsers --npmPackages --npmGlobalPackages
```

System:
    OS: macOS 10.15.5
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
    Memory: 28.13 MB / 32.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 14.15.1 - ~/.nvm/versions/node/v14.15.1/bin/node
    Yarn: 1.17.0 - /usr/local/bin/yarn
    npm: 6.14.8 - ~/.nvm/versions/node/v14.15.1/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Browsers:
    Chrome: 87.0.4280.88
    Safari: 13.1.1
  npmPackages:
    @babel/core: ^7.8.4 => 7.12.10
    @babel/runtime: ^7.8.4 => 7.12.5
    @react-native-community/eslint-config: ^1.1.0 => 1.1.0
    @react-native-community/netinfo: ^5.9.9 => 5.9.9
    amazon-cognito-identity-js: ^4.5.6 => 4.5.6
    aws-amplify: ^3.3.13 => 3.3.13
    aws-amplify-react-native: ^4.2.10 => 4.2.10
    babel-jest: ^25.1.0 => 25.5.1
    eslint: ^6.5.1 => 6.8.0
    jest: ^25.1.0 => 25.5.4
    metro-react-native-babel-preset: ^0.59.0 => 0.59.0
    react: 16.13.1 => 16.13.1
    react-native: 0.63.4 => 0.63.4
    react-native-inappbrowser-reborn: ^3.5.1 => 3.5.1
    react-native-maps: 0.27.1 => 0.27.1
    react-native-navigation: ^7.7.0 => 7.7.0
    react-test-renderer: 16.13.1 => 16.13.1
  npmGlobalPackages:
    @aws-amplify/cli: 4.36.2
    npm: 6.14.8
</details>

**Smartphone (please complete the following information):**

- Device: iPhone 11 Simulator
- Version 14.3

**Comment by erikgaas** [comment]
After some more digging I have found that the problem appears to be with whether or not there are cookies present in the browser.

If the browser has already been signed into facebook then authentication proceeds as normal. If facebook is signed out then federatedSignIn successfully signs the browser into facebook but does not successfully authenticate, therefore requiring a second attempt.

I discovered this when trying using the in app browser in ephemeral mode. Ephemeral mode causes the above error every time.

**Comment by dipeshdulal** [comment]
I have also come across this issue in iOS only. 
Looks like; 
`myapp:///?code=<code>&state=<state>#_=_` 
Removing `#_=_` and making triple slashes double slashes `//`  will authenticate. i.e.
`myapp://?code=<code>&state=<state>` 

Don't know whats happening. :cry:

**Comment by dipeshdulal** [comment]
Found a hacky solution to solve the issue to convert given url into required form using InAppBrowser as suggested https://github.com/aws-amplify/amplify-js/issues/5228 .

```js
Amplify.configure({
    ...config, oauth: {
        ...config.oauth,
        urlOpener: async function urlOpener(url, redirectUrl) {
            await InAppBrowser.isAvailable();
            const { type, url: newUrl } = await InAppBrowser.openAuth(url, redirectUrl, {
                showTitle: false,
                enableUrlBarHiding: true,
                enableDefaultShare: false,
                ephemeralWebSession: false,
            });

            const splitUrl = `myapp://?${newUrl.split("#_=_")[0].split("?")[1]}`;
            if (type === 'success') {
                Linking.openURL(splitUrl);
            }
        }

    }
})
```

**Comment by ds8k** [comment]
I'm having the exact same problem in #7737 where the URL that gets returned has three slashes in it

**Comment by wlee221** [comment]
Hello, Thanks for a detailed report and investigation! I was able to reproduce this issue on ios react native. Customers are seeing this in different issues (thanks to @ds8k for linking those):

- google / react native: #7737 
- facebook / react native / ios: #7760 

We will close them in favor of this one and communicate here. @ds8k, did you experience your issue with just ios or on android too?

**Comment by ds8k** [comment]
@wlee221 due to the way android handles accounts, I only noticed this on iOS

**Comment by Joshig99** [comment]
I have same issue on iOS, Federation social login works perfectly on android, but on iOS it requires two login attempts.  Any solution

**Comment by wlee221** [comment]
Hello, I investigated this a bit more. 

#### What does that error mean?

We use `_handleCodeFlow` to parse the redirect url query parameters and fetch the necessary tokens such as `idToken`. Due to three forward slashes,  https://github.com/aws-amplify/amplify-js/blob/7a15725f5a1c5be8bac7c055010fb0052a21bbab/packages/auth/src/OAuth/OAuth.ts#L128-L130
fails. Because we return instead of throwing, the `oauth` flow continues to execute and reaches https://github.com/aws-amplify/amplify-js/blob/7a15725f5a1c5be8bac7c055010fb0052a21bbab/packages/core/src/Credentials.ts#L417
but `idToken` is unavailable so we get the error `logins` should have length >= 1. 

#### Observations
- This only happens on ios 14 as pointed out by customers. One notable difference is that there were
```
[AuthenticationSession] Error: The provided scheme is not valid. A scheme must not include any special characters, such as ":" or "/"
[AXRuntimeCommon] Unknown client: iosFederated
```
xcode errors on ios 14 but not 13. Not sure what that first callout's `scheme` is referring to yet though. Any ideas?

- As @dipeshdulal pointed out, there's an trailing hash `#_=_` that isn't there on ios 13 or before. Why would the received url be different from ios 13?
- This happens w/ or w/out `InAppBrowser.openAuth`

#### Next Steps
Because we are not modifying the redirect url from Amplify, this is likely a problem with hosted ui. We will inquire Cognito team about this behavior. We should also `throw` early instead of `return` so that we get a better error message in future instances.

**Comment by wlee221** [comment]
> Found a hacky solution to solve the issue to convert given url into required form using InAppBrowser as suggested #5228 

Thanks a lot for finding this! This should be a good workaround while we work on this.

**Comment by derekblank** [comment]
```
[AuthenticationSession] Error: The provided scheme is not valid. A scheme must not include any special characters, such as ":" or "/"
[AXRuntimeCommon] Unknown client: iosFederated
```

> Not sure what that first callout's scheme is referring to yet though. Any ideas?

@wlee221 "Scheme" here refers to the app's URI scheme, i.e. `myapp://`.

We are also experiencing the same issue where our URI scheme is returning with three slashes: `myapp:///?code=<code>&state=<state>#_=_`

**Comment by sammartinez** [comment]
Just wanted to provide an update here. Since we do not own the Hosted UI experience from the Amplify side, we are working internally with the Amazon Cognito team on this. Once we have an update, we will provide one. Thanks for your patience!

**Comment by ahtokca** [comment]
@derekblank, wlee221 Thanks for the detailed explanation. Any news on the issue? Any safe version to rollback?

**Comment by ahtokca** [comment]
I had to downgrade to aws-amplify@3.3.11 to solve the issue

@dipeshdulal thanks for the solution, however it doesn't work for me. If I apply it then it show a warning something like 'Dear User, the app is going to do authenticate blah...'. I dont want to scary my users with extra warning :)

**Comment by magicien** [comment]
I had the same issue with aws-amplify@3.3.25

My workaround is fixing the URL in native code. (AppDelegate.m)
```
- (BOOL)application:(UIApplication *)application
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  if ([url.path isEqualToString:@"/"]) {
    NSString *fixedString = [NSString stringWithFormat:@"%@://?%@", url.scheme, url.query];
    NSURL *fixedURL = [NSURL URLWithString:fixedString];
    return [RCTLinkingManager application:application openURL:fixedURL options:options];
  }
  return [RCTLinkingManager application:application openURL:url options:options];
}
```

**Comment by sammartinez** [comment]
All, 

Wanted to provide an update here. We found that in testing that this seems to be an issue with the package we are recommending in our documentation site for [urlOpener](https://docs.amplify.aws/lib/auth/social/q/platform/js#full-samples) causing the issue. I have linked the direct issue that seems to regard this.

**Comment by manueliglesias** [comment]
**TL;DR**
I sent #8198 to addresses this by treating `'//'` as `'///'` when parsing the `currentUrl` with the `redirectSignIn`.

---

Some context:

The cognito hosted ui behaves differently if you have an active session with it or not. Assuming that you want to login with Facebook as a provider, the first time you do `federatedSignIn` you will get the Facebook UI, you'll input your facebook credentials, and then you'll have a session with cognito. The second time you do this, you already have a session with cognito, so the whole facebook ui isn't shown and you are signed in. These diagrams might do a better job explaining this:

[![image](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgT0F1dGggZmxvdywgbm8gc2Vzc2lvbiB3aXRoIENvZ25pdG8KCmFjdG9yIFVzZXIKcGFydGljaXBhbnQgQXBwAAMNACkHXG5Ib3N0ZWQgVUkgYXMgIENVUAAqDUZhY2Vib29rCgpVc2VyLT5BcHA6IHByZXNzIHNpZ24gaW4gYnV0dG9uIApub3RlIHJpZ2h0IG9mIAAkBUF1dGguZmVkZXJhdGVkU2lnbkluKFxuICB7cHJvdmlkZXI6ICcAWwgnfTtcbikKQXBwLT4rQ1VQOiBvcGVuIGJyb3dzZXIgdG8KQ1VQLT4rAIELCDogcmVkaXJlY3QgdG8AgRgHABIKSW5wdXQgY3JlZGVudGlhbHMKAIFDCC0-LQBZBQAyDABVBgCBWAUASQwAgU4FbGVmAIFNBQCBDgVlLmcuIG15YXBwOi8vLz9jb2RlPTEyMzQAgTEGAIFvBm1wbGlmeSBwYXJzZXMgdGhlIHVybCxcbmdldHMgY29kZSBhbmQgc3RhdGUAgWgGAIFoBXNlbmQADhAAgWwFLQCCbwUAGgV0b2tlbnMAglYUVXNlciBpAIMPBmVkIGlu&s=default)](https://www.websequencediagrams.com/?lz=dGl0bGUgT0F1dGggZmxvdywgbm8gc2Vzc2lvbiB3aXRoIENvZ25pdG8KCmFjdG9yIFVzZXIKcGFydGljaXBhbnQgQXBwAAMNACkHXG5Ib3N0ZWQgVUkgYXMgIENVUAAqDUZhY2Vib29rCgpVc2VyLT5BcHA6IHByZXNzIHNpZ24gaW4gYnV0dG9uIApub3RlIHJpZ2h0IG9mIAAkBUF1dGguZmVkZXJhdGVkU2lnbkluKFxuICB7cHJvdmlkZXI6ICcAWwgnfTtcbikKQXBwLT4rQ1VQOiBvcGVuIGJyb3dzZXIgdG8KQ1VQLT4rAIELCDogcmVkaXJlY3QgdG8AgRgHABIKSW5wdXQgY3JlZGVudGlhbHMKAIFDCC0-LQBZBQAyDABVBgCBWAUASQwAgU4FbGVmAIFNBQCBDgVlLmcuIG15YXBwOi8vLz9jb2RlPTEyMzQAgTEGAIFvBm1wbGlmeSBwYXJzZXMgdGhlIHVybCxcbmdldHMgY29kZSBhbmQgc3RhdGUAgWgGAIFoBXNlbmQADhAAgWwFLQCCbwUAGgV0b2tlbnMAglYUVXNlciBpAIMPBmVkIGlu&s=default)

[![image](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgT0F1dGggZmxvdywgc2Vzc2lvbiB3aXRoIGNvZ25pdG8KCmFjdG9yIFVzZXIKcGFydGljaXBhbnQgQXBwAAMNQwAqBlxuSG9zdGVkIFVJIGFzICBDVVAKClVzZXItPkFwcDogcHJlc3Mgc2lnbiBpbiBidXR0b24gCm5vdGUgcmlnaHQgb2YgACQFQXV0aC5mZWRlcmF0ZWRTaWduSW4oXG4gIHtwcm92aWRlcjogJ0ZhY2Vib29rJ307XG4pCkFwcC0-K0NVUDogb3BlbiBicm93c2VyIHRvAFMPAB4FAIErByBoYXMgYQCBdQgsXG5pbW1lZGlhdGVseSByZWRpcmVjdHMKCkNVUACBPQcADQgATwlsZWYATQplLmcuIG15YXBwOi8vP2NvZGU9MTIzNACBFwYAgVUGbXBsaWZ5IHBhcnNlcyB0aGUgdXJsLFxuZ2V0cyBjb2RlIGFuZCBzdGF0ZQCBTgYAgU4Fc2VuZAAOEACBEAUtAIJVBQAaBXRva2VucwCCPBRVc2VyIGkAgnUGZWQgaW4&s=default)](https://www.websequencediagrams.com/?lz=dGl0bGUgT0F1dGggZmxvdywgc2Vzc2lvbiB3aXRoIGNvZ25pdG8KCmFjdG9yIFVzZXIKcGFydGljaXBhbnQgQXBwAAMNQwAqBlxuSG9zdGVkIFVJIGFzICBDVVAKClVzZXItPkFwcDogcHJlc3Mgc2lnbiBpbiBidXR0b24gCm5vdGUgcmlnaHQgb2YgACQFQXV0aC5mZWRlcmF0ZWRTaWduSW4oXG4gIHtwcm92aWRlcjogJ0ZhY2Vib29rJ307XG4pCkFwcC0-K0NVUDogb3BlbiBicm93c2VyIHRvAFMPAB4FAIErByBoYXMgYQCBdQgsXG5pbW1lZGlhdGVseSByZWRpcmVjdHMKCkNVUACBPQcADQgATwlsZWYATQplLmcuIG15YXBwOi8vP2NvZGU9MTIzNACBFwYAgVUGbXBsaWZ5IHBhcnNlcyB0aGUgdXJsLFxuZ2V0cyBjb2RlIGFuZCBzdGF0ZQCBTgYAgU4Fc2VuZAAOEACBEAUtAIJVBQAaBXRva2VucwCCPBRVc2VyIGkAgnUGZWQgaW4&s=default)

I intercepted the https traffic to inspect the request/responses and validated what was mentioned before: on an initial login (when no active session exists on the cognito hosted ui), the `currentUrl` that RN reports when coming back from all the redirects, has three slashes; the next time it comes back with just two: (I don't have a good explanation as of why this happens)

**Initial sign in**
![image](https://user-images.githubusercontent.com/6154160/116474515-767f9700-a82d-11eb-8e6f-5ae8b39c4809.png)
<img width="1897" alt="image" src="https://user-images.githubusercontent.com/6154160/116474529-797a8780-a82d-11eb-9c0f-ca046b99ede3.png">

---

**Subsequent sign in**
![image](https://user-images.githubusercontent.com/6154160/116474542-7e3f3b80-a82d-11eb-8a63-a43df0cca98d.png)
<img width="1888" alt="image" src="https://user-images.githubusercontent.com/6154160/116474555-8303ef80-a82d-11eb-88ad-dfd5e3b9697c.png">

---

As pointed out by @ahtokca, this code doesn't handle well the three slashes scenario:

https://github.com/clethrill/amplify-js/blob/d3e522b812d4e0b52dd88b5a9ad83ebd68ad8b56/packages/auth/src/OAuth/OAuth.ts#L128-L130
```typescript
		if (!code || parse(currentUrl).pathname !== parse(this._config.redirectSignIn).pathname) {
			return;
		}
```

Node's `url`'s `parse` function returns `null` for `pathname` when the url has two slashes after the scheme:

```diff
❯ node
Welcome to Node.js v14.15.1.
Type ".help" for more information.
❯ const { parse } = require('url');
undefined
❯ parse('myapp:///?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr#')
Url {
  protocol: 'myapp:',
  slashes: true,
  auth: null,
  host: '',
  port: null,
  hostname: '',
  hash: '#',
  search: '?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr',
  query: 'code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr',
+  pathname: '/',
  path: '/?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr',
  href: 'myapp:///?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr#'
}
❯ parse('myapp://?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr#')
Url {
  protocol: 'myapp:',
  slashes: true,
  auth: null,
  host: '',
  port: null,
  hostname: '',
  hash: '#',
  search: '?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr',
  query: 'code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr',
+  pathname: null,
  path: '?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr',
  href: 'myapp://?code=508c2bb4-98a8-4d5c-9d59-169faa819524&state=tuwIPFegcrL7KJrijhQe1oAGC6eLWFdr#'
}
```

My PR just assumes that if there are only two slashes, the pathname will be `'/'`, this seems to be aligned with: https://github.com/expo/expo/issues/6497#issuecomment-574882448

If everything looks good in the PR, you might be able to try this on unstable soon.

**Comment by ahtokca** [comment]
Wow, @manueliglesias, thanks a lot. Looking forward to having your PR merged :)

**Comment by tlmader** [comment]
I ran into the below error last night, and my redirect did not contain a trailing slash.
```
validation error detected: Value '{cognito-idp.us-east-1.amazonaws.com/us-east-xxx=}' at 'logins' failed to satisfy constraint: Map value must satisfy constraint: [Member must have length less than or equal to 50000, Member must have length greater than or equal to 1]"}
```
My issue was that I needed to add `responseType: 'code'` to the Amplify `oauth` configuration object.

My Amplify configuration:
```
const awsconfig = {
  Auth: {
    userPoolId: 'us-east-1_x',xxx
    userPoolWebClientId: 'xxx',
    identityPoolId: 'us-east-1:xxx',
    region: 'us-east-1',
    oauth: {
      domain: 'xxx.auth.us-east-1.amazoncognito.com',
      redirectSignIn: redirectUri,
      redirectSignOut: redirectUri,
      responseType: 'code',
      urlOpener,
    },
  },
};
```
What helped me to figure out the exact problem was adding `window.LOG_LEVEL = 'DEBUG';` and I was able to get the following message:
```
[DEBUG] 57:07.955 OAuth - Starting undefined flow with exp://127.0.0.1:19000?code=xxx
```
Hope this helps someone!

**Comment by tlmader** [comment]
I ran into the below error last night, and my redirect did not contain a trailing slash.
```
validation error detected: Value '{cognito-idp.us-east-1.amazonaws.com/us-east-xxx=}' at 'logins' failed to satisfy constraint: Map value must satisfy constraint: [Member must have length less than or equal to 50000, Member must have length greater than or equal to 1]"}
```
My issue was that I needed to add `responseType: 'code'` to the Amplify `oauth` configuration object.

My Amplify configuration:
```
const awsconfig = {
  Auth: {
    userPoolId: 'us-east-1_x',xxx
    userPoolWebClientId: 'xxx',
    identityPoolId: 'us-east-1:xxx',
    region: 'us-east-1',
    oauth: {
      domain: 'xxx.auth.us-east-1.amazoncognito.com',
      redirectSignIn: redirectUri,
      redirectSignOut: redirectUri,
      responseType: 'code',
      urlOpener,
    },
  },
};
```
What helped me figure out the exact problem was adding `window.LOG_LEVEL = 'DEBUG';` and I was able to get the following message:
```
[DEBUG] 57:07.955 OAuth - Starting undefined flow with exp://127.0.0.1:19000?code=xxx
```
Hope this helps someone!
